Java Collections Framework: Set

fvaray opened the pull request.

fvaray committed:
Java Collections Framework: Set

Задачи 1,2

AleksandrZhukovJava commented on src/org/skypro/skyshop/search/SearchEngine.java:
```diff
@@ -6,11 +6,11 @@
 public class SearchEngine {
     //private final Map<String, Searchable> searchable = new HashMap<>();
-    private final List<Searchable> searchable = new LinkedList<>();
+    private final Set<Searchable> searchable = new LinkedHashSet<>();
```
Не вижу необходимости в сохранении порядка вставки тут, можно сэкономить ресурсы и взять HashSet

AleksandrZhukovJava commented on src/org/skypro/skyshop/search/SearchEngine.java:
```diff
@@ -36,4 +36,17 @@ public Map<String, Integer> searchRepeat(String term) throws BestResultNotFound
+        public int compare(String s1, String s2) {
+            return s2.length() - s1.length();
+        }
+    }
```
1) нужно убрать в отдельный класс
2) сортировка нужна сложнее, по условиям это длинна, а потом в натуральном порядке, даже если работает по умолчанию, лучше определить это явно

fvaray committed:
Java Collections Framework: Set

Исправил замечания наставника.

fvaray committed:
Java Collections Framework: Set

Исправил замечания наставника.

fvaray committed:
Java Collections Framework: Set

Исправил замечания наставника. Подчистил мейн.

AleksandrZhukovJava commented on src/org/skypro/skyshop/search/ReverseStringComparator.java:
```diff
@@ -0,0 +1,18 @@
+
+public class ReverseStringComparator implements Comparator<String> {
+    @Override
+    public int compare(String s2, String s1) {
+        int compareResult;
+        compareResult = Integer.compare(s1.length(), s2.length());
+        if (compareResult == 0) {
+            compareResult = s2.compareTo(s1);
+            return compareResult;
+        }
+        else{
+            return compareResult;
+        }
+    }
```
Тут основная проблема что компаратор должен сравнивать не стринги а объекты Searchable, нам нужен именно способ сравнивать объекты

AleksandrZhukovJava commented on src/org/skypro/skyshop/search/SearchEngine.java:
```diff
@@ -5,12 +5,11 @@
+    private final Set<Searchable> searchable = new HashSet<>();
 
-    public Map<String, Integer> searchRepeat(String term) throws BestResultNotFound {
+    public Set searchRepeat(String term) throws BestResultNotFound {
```
Нужно добавить дженерики

AleksandrZhukovJava commented on src/org/skypro/skyshop/search/SearchEngine.java:
```diff
@@ -5,12 +5,11 @@
+    public Set searchRepeat(String term) throws BestResultNotFound {
         int quantity = 0;
-        Map<String,Integer> searchableList = new TreeMap<>();
+        Set <String> searchableList = new TreeSet<>(new ReverseStringComparator());
```
Должен возвращать Searchable как дженерик сэта

fvaray replied:
Александр, я когда вместо String добавляю Searchable, у меня сразу начинает ругаться компилятор на TreeSet

fvaray committed:
Java Collections Framework: Set

Исправил замечания наставника. Добавил в метод сеарч дженерик. копмаратор сранивает не по стрингам а по объектам.
